diff --git a/src/d3d9/meson.build b/src/d3d9/meson.build
index 2f45946..98954c5 100644
--- a/src/d3d9/meson.build
+++ b/src/d3d9/meson.build
@@ -52,7 +52,7 @@ d3d9_src = [
 d3d9_ld_args      = []
 d3d9_link_depends = []
 
-if platform != 'windows'
+if platform != 'windows' and platform != 'darwin'
   d3d9_ld_args      += [ '-Wl,--version-script', join_paths(meson.current_source_dir(), 'd3d9.sym') ]
   d3d9_link_depends += files('d3d9.sym')
 endif
diff --git a/src/dxvk/dxvk_adapter.cpp b/src/dxvk/dxvk_adapter.cpp
index d890765..83fd814 100644
--- a/src/dxvk/dxvk_adapter.cpp
+++ b/src/dxvk/dxvk_adapter.cpp
@@ -168,6 +168,26 @@ namespace dxvk {
     std::vector<VkExtensionProperties> extensions(extensionCount);
     m_capabilities.queryDeviceExtensions(&extensionCount, extensions.data());
 
+#ifdef __APPLE__
+    uint32_t supportedExtensionCount = 0u;
+    vk->vkEnumerateDeviceExtensionProperties(
+      m_handle, nullptr, &supportedExtensionCount, nullptr);
+
+    std::vector<VkExtensionProperties> supportedExtensions(supportedExtensionCount);
+    vk->vkEnumerateDeviceExtensionProperties(
+      m_handle, nullptr, &supportedExtensionCount, supportedExtensions.data());
+
+    auto portabilitySubset = std::find_if(
+      supportedExtensions.begin(), supportedExtensions.end(),
+      [] (const VkExtensionProperties& extension) {
+        return !std::strcmp(
+          extension.extensionName, "VK_KHR_portability_subset");
+      });
+
+    if (portabilitySubset != supportedExtensions.end())
+      extensions.push_back(*portabilitySubset);
+#endif
+
     for (const auto& extra : m_extraExtensions) {
       bool found = false;
 
diff --git a/src/dxvk/dxvk_instance.cpp b/src/dxvk/dxvk_instance.cpp
index 541caaf..56e8e0b 100644
--- a/src/dxvk/dxvk_instance.cpp
+++ b/src/dxvk/dxvk_instance.cpp
@@ -209,6 +209,16 @@ namespace dxvk {
         extensionsEnabled.insert(ext);
     }
 
+#ifdef __APPLE__
+    auto portabilityExtension = extensionsSupported.find(
+      vk::makeExtension(VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME));
+
+    bool enablePortabilityEnumeration = portabilityExtension != extensionsSupported.end();
+
+    if (enablePortabilityEnumeration)
+      extensionsEnabled.insert(*portabilityExtension);
+#endif
+
     Logger::info("Enabled instance extensions:");
 
     for (const auto& ext : extensionsEnabled) {
@@ -239,6 +249,10 @@ namespace dxvk {
       appInfo.apiVersion            = DxvkVulkanApiVersion;
 
       VkInstanceCreateInfo info = { VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO };
+#ifdef __APPLE__
+      if (enablePortabilityEnumeration)
+        info.flags |= VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR;
+#endif
       info.pApplicationInfo         = &appInfo;
       info.enabledLayerCount        = layerNames.size();
       info.ppEnabledLayerNames      = layerNames.data();
diff --git a/src/util/util_env.cpp b/src/util/util_env.cpp
index c86c3ab..d09db13 100644
--- a/src/util/util_env.cpp
+++ b/src/util/util_env.cpp
@@ -6,6 +6,8 @@
 #ifdef __linux__
 #include <unistd.h>
 #include <limits.h>
+#elif defined(__APPLE__)
+#include <mach-o/dyld.h>
 #elif defined(__FreeBSD__)
 #include <sys/sysctl.h>
 #include <unistd.h>
@@ -93,6 +95,15 @@ namespace dxvk::env {
     size_t count = readlink("/proc/self/exe", exePath.data(), exePath.size());
 
     return std::string(exePath.begin(), exePath.begin() + count);
+#elif defined(__APPLE__)
+    uint32_t size = 0;
+    _NSGetExecutablePath(nullptr, &size);
+
+    std::vector<char> exePath(size);
+    if (_NSGetExecutablePath(exePath.data(), &size) != 0)
+      return "";
+
+    return std::filesystem::weakly_canonical(exePath.data()).string();
 #elif defined(__FreeBSD__)
     int mib[4] = {CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, getpid()};
     char exePath[PATH_MAX] = {};
@@ -127,7 +138,11 @@ namespace dxvk::env {
 #else
     std::array<char, 16> posixName = {};
     dxvk::str::strlcpy(posixName.data(), name.c_str(), 16);
+#ifdef __APPLE__
+    ::pthread_setname_np(posixName.data());
+#else
     ::pthread_setname_np(pthread_self(), posixName.data());
+#endif
 #endif
   }
 
diff --git a/src/util/util_math.h b/src/util/util_math.h
index 9276d93..baff3ce 100644
--- a/src/util/util_math.h
+++ b/src/util/util_math.h
@@ -1,10 +1,11 @@
 #pragma once
 
+#include <cstddef>
 #include <cmath>
 
 namespace dxvk {
   
-  constexpr size_t CACHE_LINE_SIZE = 64;
+  constexpr std::size_t CACHE_LINE_SIZE = 64;
   constexpr double pi = 3.14159265359;
 
   template<typename T>
diff --git a/src/util/util_small_vector.h b/src/util/util_small_vector.h
index 861c9e0..51e7b26 100644
--- a/src/util/util_small_vector.h
+++ b/src/util/util_small_vector.h
@@ -194,7 +194,7 @@ namespace dxvk {
 
     size_t pick_capacity(size_t n) {
       // Pick next largest power of two for the new capacity
-      return size_t(1u) << ((sizeof(n) * 8u) - bit::lzcnt(n - 1));
+      return size_t(1u) << ((sizeof(n) * 8u) - bit::lzcnt(uint64_t(n - 1)));
     }
 
     T* ptr(size_t idx) {
diff --git a/src/util/util_win32_compat.h b/src/util/util_win32_compat.h
index 5b156c6..c9b8ac1 100644
--- a/src/util/util_win32_compat.h
+++ b/src/util/util_win32_compat.h
@@ -1,6 +1,6 @@
 #pragma once
 
-#if defined(__unix__)
+#if defined(__unix__) || defined(__APPLE__)
 
 #include <windows.h>
 #include <dlfcn.h>
diff --git a/src/vulkan/vulkan_loader.cpp b/src/vulkan/vulkan_loader.cpp
index f9d1e98..fc5da18 100644
--- a/src/vulkan/vulkan_loader.cpp
+++ b/src/vulkan/vulkan_loader.cpp
@@ -14,6 +14,9 @@ namespace dxvk::vk {
 #ifdef _WIN32
       "winevulkan.dll",
       "vulkan-1.dll",
+#elif defined(__APPLE__)
+      "libMoltenVK.dylib",
+      "libvulkan.dylib",
 #else
       "libvulkan.so",
       "libvulkan.so.1",
